### ISSUE TITLE
Question about training

### ISSUE THREAD

**Issue by RuixiangXue** [YihangChen-ee/FCGS#11]
Hi, Chen.
Thanks for your kind sharing of code. I’m trying to reproduce the FCGS training process. 
However, I found that at the beginning of training, the 'mask_feq' is entirely False, which causes cuda errors later in the pipeline. Have you encountered this issue before? If so, how did you handle or avoid it?

**Comment by RuixiangXue** [comment]
I’d really appreciate it if I could also get some help with a few other issues I’ve been running into:
1. What is the initial learning rate, optimizer, and lr_scheduler?
2. During training, is the entire Gaussian model processed in a single iteration? Or do you also use the chunking strategy from the testing code? I noticed that the GPU memory consumption is quite high, so I’m wondering how you handled that during training.
3. When calculating the fidelity loss, do you include all the test views?

**Comment by YihangChen-ee** [comment]
Hi Ruixiang,

Here are some tips you might need for training:

1. The masking threshold is small so that the model tends to assign all the masks as 1 in the initialization.
2. If you directly train the entire model from scratch, the masks would soon collapse to all 0s since the MLPs in mask=1 path are random values, and the fastest way for the model to converge is setting mask as all 0s.
3. However, 2 is not what we want. Therefore, the MLPs in mask=1 path should be sufficiently pre-trained to maintain fidelity before other MLPs are involved. We have discussed this phenomenon in our paper implementation details. To train the MLPs in the mask=1 path in the first stage, you could consider such a loss: L=L_fidelity + lambda_m * (-mask_ratio) to sufficiently train the MLPs in mask=1 path and the mask MLP as well. Initially, you could assign a larger lambda_m value (eg, 3e-3) to encourage more 1s in the mask, which helps with the learning of MLPs in this path. With the training process goes, you could reduce lambda_m value (eg, to 5e-4) to find a good balance between the mask ratio and the fidelity.
6. After these MLPs in the m=1 path are sufficiently trained in the first stage, you could involve the entropy constraint to train the entire model, and gradually reduce the lambda_m value until this item lambda_m * (-mask_ratio) is totally removed.

For your questions:
1. In the first stage (ie, only train the MLPs in m=1 path and train the mask MLP), initial lr is 2e-4. We reduce the lr by half after the model is converged at the current lr. For the second stage, initial lr is 0.5e-4, and the same lr adjustment.
2. We sample 1500k Gaussians for each scene when training the entire model. In the first stage, I recommend you sample fewer Gaussians initially, since the quantization steps are not well learned initially, so that the quantized Gaussian attributes are of large deviations, which can result in abnormally values of opacity or scale or rotation, making the GPU memory consumption high.
3. The training process is run only on the training set. In the training set, 4 training views and 2 testing views are randomly sampled for each scene when training the model.

Best regards,
Yihang

**Comment by anonymouscvsubmission** [comment]
> 2. We sample 1500k Gaussians for each scene when training the entire model. In the first stage, I recommend you sample fewer Gaussians initially, since the quantization steps are not well learned initially, so that the quantized Gaussian attributes are of large deviations, which can result in abnormal values of opacity or scale or rotation, making the rendering process less efficient, and the GPU memory consumption high.
Sir, may I ask, if part of the Gaussians are sampled for training, how the supervision is calculated through rendering? 
Part of the Gaussians yields incomplete rendering result, how do you construct the loss between **Ground Truth image** and the **incomplete rendering result**?

**Comment by YihangChen-ee** [comment]
Thanks for your reminder! I misremembered this part. Sampling 1500k Gaussians for each scene was our initial approach, but we found that it did not work well. Regarding your question, we will merge the remaining unsampled Gaussian points with the decoded 1500k Gaussian points for rendering, so we can still obtain a complete rendered image. 
What we do eventually is that we only used scenes with less than 1500k Gaussian points for training (all the Gaussians are used in the scene), and the number of such scenes is still sufficient.

**Comment by RuixiangXue** [comment]
Thank you so much for your thorough and detailed explanation. I’ll revise my code based on your kind suggestions.
Regarding the sampling strategy during training, could you share what kind of method you used?

**Comment by YihangChen-ee** [comment]
> Thank you so much for your thorough and detailed explanation. I’ll revise my code based on your kind suggestions. Regarding the sampling strategy during training, could you share what kind of method you used?

Hi, I have rephrased some sentences in my initial response. Please check the revised response.

For the sampling strategy, you could use 'torch.randperm(g_xyz.shape[0])[:N]' for sampling. After the sampled Gaussians are encoded and decoded in the model, you can combine them with the unsampled ones for a complete image rendering for supervision.

**Comment by RuixiangXue** [comment]
I modified my training code. Now the GPU memory usage during training has been reduced, but each training iteration still takes quite a long time (around 2 minutes). I’d like to ask how long a single iteration typically takes in your case. I’m wondering if the data loading process might be the bottleneck. Could you share what your data consists of during loading?
